src/Instrumentation/ExtAmqp/src/ExtAmqpInstrumentation.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\ExtAmqp;
+
+use AMQPExchange;
+use AMQPQueue;
+use OpenTelemetry\API\Globals;
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\API\Trace\StatusCode;
+use OpenTelemetry\Context\Context;
+use OpenTelemetry\Context\Propagation\ArrayAccessGetterSetter;
+use function OpenTelemetry\Instrumentation\hook;
+use OpenTelemetry\SemConv\TraceAttributes;
+use Throwable;
+
+final class ExtAmqpInstrumentation
+{
+    public const NAME = 'ext_amqp';
+
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.ext_amqp');
+
+        hook(
+            AMQPExchange::class,
+            'publish',
+            pre: static function (
+                AMQPExchange $channel,
+                array $params,
+                string $class,
+                string $function,
+                ?string $filename,
+                ?int $lineno,
+            ) use ($instrumentation): array {
+                $routingKey = $params[1];
+
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = $instrumentation
+                    ->tracer()
+                    ->spanBuilder($routingKey . ' publish')
+                    ->setSpanKind(SpanKind::KIND_PRODUCER)
+                    // code
+                    ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
+                    ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
+                    ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
+                    ->setAttribute(TraceAttributes::CODE_LINENO, $lineno)
+                    // messaging
+                    ->setAttribute(TraceAttributes::MESSAGING_SYSTEM, 'rabbitmq')
```

Comment:
technically, it will work with any amqp broker. good catch. Let me see if I can use some of the available context to find out.

Reply:
I think we can't know the broker type from ext-amqp - it could be anything, really. I propose we switch from `rabbitmq` to `amqp` on this required attribute